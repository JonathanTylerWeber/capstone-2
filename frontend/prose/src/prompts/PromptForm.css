pre {
  color: white;
}

p {
  color: white;
}

.form {
  margin-top: 10em;
}