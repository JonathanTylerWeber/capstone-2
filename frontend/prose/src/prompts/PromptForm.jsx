import React, { useState } from 'react';
import ProseApi from '../API/api';
import { Container, Form, Button, Spinner } from "react-bootstrap"
import "./PromptForm.css"

function PromptForm() {
  const [type, setType] = useState('');
  const [adj, setAdj] = useState('');
  const [prompt, setPrompt] = useState('');
  const [ratingResult, setRatingResult] = useState('');
  const [rewriteResult, setRewriteResult] = useState('');
  const [isLoading, setIsLoading] = useState(false);

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      setIsLoading(true);
      const ratingResponse = await ProseApi.getRating(type, adj, prompt);
      const rewriteResponse = await ProseApi.rewritePrompt(type, adj, prompt);
      console.log('Rating response for form:', ratingResponse);
      console.log('Rewrite response for form:', rewriteResponse);
      setRatingResult(ratingResponse);
      setRewriteResult(rewriteResponse);
      await ProseApi.savePrompt(ratingResponse, prompt, rewriteResponse, type, adj)
    } catch (error) {
      console.error("Failed to submit prompt form:", error);
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <Container className="m-auto form-cntr ">
      <Form onSubmit={handleSubmit} className='form'>
        <h1 className="form-h1">Enter writing and info to get rating, tips, and a rewrite</h1>
        <Form.Group className="mb-3" controlId="type">
          <Form.Label>Type:</Form.Label>
          <Form.Control type="text" value={type}
            onChange={(e) => setType(e.target.value)}
            required placeholder="e.g. 'thank you letter', 'product description', etc" />
        </Form.Group>
        <Form.Group className="mb-3" controlId="adj">
          <Form.Label>Adjective:</Form.Label>
          <Form.Control type="text" value={adj}
            onChange={(e) => setAdj(e.target.value)}
            required placeholder="e.g. 'sincere', 'inspirational', etc" />
        </Form.Group>
        <Form.Group className="mb-3" controlId="prompt">
          <Form.Label>Prompt:</Form.Label>
          <Form.Control as="textarea" value={prompt}
            onChange={(e) => setPrompt(e.target.value)}
            required placeholder="Enter your writing to get feedback"
            rows={6} />
        </Form.Group>
        <Button className="btn form-btn" variant="primary" type="submit">
          {isLoading ? (
            <>
              <Spinner animation="border" size="sm" />
              <span className="ms-2" style={{ color: 'black' }} >Submitting...</span>
            </>
          ) : (
            "Submit"
          )}
        </Button>
      </Form>
      {ratingResult &&
        <div className='response'>
          <h2>Rating:</h2>
          <p>{ratingResult}</p>
          <br />
          <h2>Rewritten {type}:</h2>
          <p>{rewriteResult}</p>
        </div>
      }
    </Container>
  );
}

export default PromptForm;
