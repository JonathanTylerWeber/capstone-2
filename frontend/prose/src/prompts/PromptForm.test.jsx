import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import PromptForm from './PromptForm';
import fetchMock from 'fetch-mock';

describe('PromptForm', () => {
  // Mock the rating and rewrite responses
  const mockRatingResponse = 'Your rating response';
  const mockRewriteResponse = 'Your rewrite response';

  beforeEach(() => {
    // Reset fetch mock before each test
    fetchMock.reset();
  });

  it('renders without crashing', async () => {
    render(<PromptForm />);
    expect(screen.getByText('Enter writing and info to get rating, tips, and a rewrite')).toBeInTheDocument();
  });

  it('displays loading spinner when submitting form', async () => {
    render(<PromptForm />);
    const typeInput = screen.getByPlaceholderText("e.g. 'thank you letter', 'product description', etc");
    const adjInput = screen.getByPlaceholderText("e.g. 'sincere', 'inspirational', etc");
    const promptInput = screen.getByPlaceholderText("Enter your writing to get feedback");

    fireEvent.change(typeInput, { target: { value: 'Thank You Letter' } });
    fireEvent.change(adjInput, { target: { value: 'Sincere' } });
    fireEvent.change(promptInput, { target: { value: 'Thank you for your help.' } });

    const submitButton = screen.getByText('Submit');
    fireEvent.click(submitButton);
    await waitFor(() => expect(screen.getByText('Submitting...')).toBeInTheDocument());
  });

  it('displays rating result and rewritten prompt after submission', async () => {
    // Mock the API responses
    fetchMock.post('/submit/rating', { choices: [{ message: { content: mockRatingResponse } }] });
    fetchMock.post('/submit/rewrite', { choices: [{ message: { content: mockRewriteResponse } }] });

    render(<PromptForm />);
    const typeInput = screen.getByPlaceholderText("e.g. 'thank you letter', 'product description', etc");
    const adjInput = screen.getByPlaceholderText("e.g. 'sincere', 'inspirational', etc");
    const promptInput = screen.getByPlaceholderText("Enter your writing to get feedback");

    fireEvent.change(typeInput, { target: { value: 'Thank You Letter' } });
    fireEvent.change(adjInput, { target: { value: 'Sincere' } });
    fireEvent.change(promptInput, { target: { value: 'Thank you for your help.' } });

    const submitButton = screen.getByText('Submit');
    fireEvent.click(submitButton);

    // Wait for the response elements to appear
    await waitFor(() => {
      expect(screen.getByText(/Your rating response/i)).toBeInTheDocument();
      expect(screen.getByText(/Your rewrite response/i)).toBeInTheDocument();
    });
  });
});
