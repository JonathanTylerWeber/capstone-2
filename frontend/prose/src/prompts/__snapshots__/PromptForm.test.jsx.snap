// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PromptForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="m-auto form-cntr  container"
  >
    <form
      class="form"
    >
      <h1
        class="form-h1"
      >
        Enter writing and info to get rating, tips, and a rewrite
      </h1>
      <div
        class="mb-3"
      >
        <label
          class="label form-label"
          for="type"
        >
          Type:
        </label>
        <input
          class="form-control"
          id="type"
          placeholder="e.g. 'thank you letter', 'product description', etc"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="label form-label"
          for="adj"
        >
          Adjective:
        </label>
        <input
          class="form-control"
          id="adj"
          placeholder="e.g. 'sincere', 'inspirational', etc"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="label form-label"
          for="prompt"
        >
          Prompt:
        </label>
        <textarea
          class="form-control"
          id="prompt"
          placeholder="Enter your writing to get feedback"
          required=""
          rows="6"
        />
      </div>
      <button
        class="btn form-btn btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
